### PR TITLE
Revamp the viewer to get rid of javaFX and add new features

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-viewer/src/main/java/org/sonar/java/ast/SyntaxTreeDebug.java [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+
+public class SyntaxTreeDebug {
+
+  private SyntaxTreeDebug() {
+  }
+
+  public static String toString(Tree syntaxNode) {
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/ast/SyntaxTreeDebug.java]: MAJOR This method has 111 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)

File: java-viewer/src/main/java/org/sonar/java/cfg/CFGDebug.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    buffer.append(cfg.entry().id());
+    buffer.append('\n');
+    buffer.append('\n');
+    for (Block block : cfg.blocks()) {
+      buffer.append(toString(block));
+    }
+    return buffer.toString();
+  }
+
+  private static String toString(CFG.Block block) {
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/cfg/CFGDebug.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

File: java-viewer/src/main/java/org/sonar/java/cfg/CFGDebug.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    buffer.append('\n');
+    buffer.append('\n');
+    return buffer.toString();
+  }
+
+  private static void appendKind(StringBuilder buffer, Kind kind) {
+    String name = kind.name();
+    int n = MAX_KINDNAME - name.length();
+    buffer.append(name);
+    while (--n >= 0) {
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/cfg/CFGDebug.java]: MAJOR Extract this increment or decrement operator into a dedicated statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS881)

File: java-viewer/src/main/java/org/sonar/java/se/EGDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import javax.annotation.Nullable;
+
+import java.text.MessageFormat;
+
+public class EGDataProvider {
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/se/EGDataProvider.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: java-viewer/src/main/java/org/sonar/java/viewer/DotHelper.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    TOKEN_KIND("tokenKind"),
+    CLASS_KIND("classKind"),
+    METHOD_KIND("methodKind"),
+
+    EXCEPTION_EDGE("exceptionEdge"),
+    YIELD_EDGE("yieldEdge");
+
+    private final String name;
+
+    private Highlighting(String name) {
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/viewer/DotHelper.java]: MINOR "private" is redundant in this context. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2333)

File: java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  private static String renderWithValues(String javaCode, Map<String, String> values) {
+    values.put("javaCode", javaCode);
+    return new VelocityTemplateEngine().render(new ModelAndView(values, "velocity/index.vm"));
+  }
+
+  private static String defaultFileContent() {
+    String result;
+    try {
+      Path path = Paths.get(Viewer.class.getResource(DEFAULT_SOURCE_CODE_LOCATION).toURI());
+      result = new String(Files.readAllBytes(path));
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java]: MAJOR Remove this use of constructor "String(byte[])" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1943)

File: java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+public class Viewer {
+
+  private static final Logger LOGGER = Logger.getLogger(Viewer.class);
+  private static final String DEFAULT_SOURCE_CODE_LOCATION = "/public/example/example.java";
+  private static final String DEFAULT_SOURCE_CODE = defaultFileContent();
+
+  public static void main(String[] args) {
+    BasicConfigurator.configure();
+
+    exception(Exception.class, (e, req, res) -> LOGGER.error("Unexpected exception.", e)); // print all exceptions
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-viewer/src/main/java/org/sonar/java/cfg/CFGViewer.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+public class CFGViewer {
+
+  private CFGViewer() {
+  }
+
+  private static final String DOT_PADDING = "  ";
```
Comment (SonarTech) [inline java-viewer/src/main/java/org/sonar/java/cfg/CFGViewer.java]: MAJOR Remove this unused "DOT_PADDING" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-viewer/src/main/resources/public/lib/viewer/viewer.js [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+      }
+      result[details.ppKey].push(node.id);
+    }
+  });
+  return result;
+}
+
+function mapPPByLine(cfgCode) {
+  let result = {};
+  let currentBlock;
```
Comment (SonarTech) [inline java-viewer/src/main/resources/public/lib/viewer/viewer.js]: MAJOR Move the declaration of "currentBlock" to line 394. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=javascript%3AS2392)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by benzonico** [comment]
As discussed, there is quite some refactoring and simplification to be done on java code before we can merge this PR. Please also add the minimal required documentation about how this can be run.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java [hunk 10]
```diff
@@ -25,40 +25,44 @@
...
+  public abstract String name();
+
+  /**
+   * Fill nodes and edges for the graph
+   * @return the completed graph
+   */
+  public abstract DotGraph build();
+
+  /**
+   * Convert the CFG to DOT format (graph description language).
```
Comment (benzonico) [inline java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java]: wrong comment

File: java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java [hunk 11]
```diff
@@ -25,40 +25,44 @@
...
+   * See language specification: http://www.graphviz.org/content/dot-language
+   */
+  public final String toDot() {
+    build();
+
+    StringBuilder sb = new StringBuilder("graph " + name() + "{");
+    for (DotGraph.Node node : nodes) {
```
Comment (benzonico) [inline java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java]: matter of taste I would go for stream api here nodes.map(Node::toDot).forEach(sb::append) ? so you'll end up with two very similar lines for node and edges.

File: java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java [hunk 12]
```diff
@@ -25,40 +25,44 @@
...
+  public abstract DotGraph build();
+
+  /**
+   * Convert the CFG to DOT format (graph description language).
+   * See language specification: http://www.graphviz.org/content/dot-language
+   */
+  public final String toDot() {
+    build();
+
+    StringBuilder sb = new StringBuilder("graph " + name() + "{");
```
Comment (benzonico) [inline java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java]: if you use a builder, use append

File: java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java [hunk 13]
```diff
@@ -88,20 +92,55 @@ public static String name(@Nullable Highlighting highlighting) {
     }
   }
 
-  public abstract static class Node extends DotDataProvider {
+  private abstract static class DotElement {
```
Comment (benzonico) [inline java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java]: Not found of this factorization by inheritance here. This seems rather artificial and seems to complicate things a bit.

File: java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java [hunk 14]
```diff
@@ -83,20 +95,32 @@ private static String generate(Request request) {
...
 
     return renderWithValues(javaCode, values);
   }
 
+  private static MethodTree getFirstMethod(CompilationUnitTree cut) {
+    ClassTree classTree = (ClassTree) cut.types().get(0);
+    return (MethodTree) classTree.members().stream()
+      .filter(m -> m.is(Tree.Kind.METHOD))
+      .findFirst()
+      .orElse(null);
```
Comment (benzonico) [inline java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java]: given that this method is not annotated and the null value is passed to CFG.build without null check,  this means this is not unit tested AND this should problaby be  null checked to return a nice error.

Comment (m-g-sonar) [inline java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java]: handled with a nice error displayed in the UI. Not unit tested.

File: java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java [hunk 15]
```diff
@@ -88,15 +89,22 @@ public static void main(String[] args) {
...
   private static String generate(Request request) {
     String javaCode = Optional.ofNullable(request.queryParams("javaCode")).orElse(DEFAULT_SOURCE_CODE);
+    return generate(javaCode);
+  }
+
+  private static String generate(String javaCode) {
+    Map<String, String> values;
     try {
-      return renderIndex(javaCode);
+      values= getValues(javaCode);
```
Comment (benzonico) [inline java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java]: formatting

File: java-viewer/src/main/resources/public/lib/viewer/viewer.js (hunk 9 above)
Comment (m-g-sonar) [inline java-viewer/src/main/resources/public/lib/viewer/viewer.js]: I need `currentBlock` to be outside the `for` loop as I'll compute it when encountered on certain lines, and then use it for next ones.

File: java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java (hunk 14 above)
Comment (m-g-sonar) [inline java-viewer/src/main/java/org/sonar/java/viewer/Viewer.java]: Added dedicated unit tests to cover these cases.

**Pull request merged**